### DISCUSSION TITLE
Make the trpc API simple

### DISCUSSION THREAD

**Discussion by cloudcompute** [trpc/trpc#1537] [Ideas]
A lot of boilerplate code is there in the current implementation:

a. Method chaining
b. create router
c. ...

Consider this TS interface
```
interface User {
  name: string;
  email: string;
  ...
}
```

Let the users create the two interfaces where the ```trpcQueries``` and ```trpcMutations``` are standard ```trpc``` keywords. For example:

```
interface trpcQueries {
  getUser: () => User;
  getUsers: () => User[];
  ...
}

interface trpcMutations {
  createUser: (user: User) => boolean;
  ...
}
```

```createRouter()```, ```.query()```, ```.mutation()```, ```resolve()```, etc. is  all boilerplate. The users will simply implement the ```trpcQueries``` and ```trpcMutations``` interfaces.

```
const createUser = (user: User) : boolean => { 
  ...
}
```

This design has a disadvantage though. The user needs to maintain the method signatures at two places. A much better alternative is to let us use the ```trpcQuery``` and ```trpcMutation``` decorators.

```
@trpcMutation
const createUser = (user: User) : boolean => { ... }
```

As it is mentioned in the trpc docs.. ```There's no internal difference between queries and mutations apart from semantics.```, a single decorator ```trpc``` may do.
```
@trpc
const createUser = (user: User) : boolean => { ... }
```
Currently the user is resolving the input using the different validators. Looks like the ```reflect-metadata``` package might help us resolve the method's signatures.

dear community, kindly give your suggestions.

Thanks

**Comment by KATT** [comment]
I'm not going to switch to enforcing output schemas, that sort of ruins the point of tRPC. That said, I am looking at ways to simplify this stuff and to allow to more fun DX-things, if you're interested in getting involved, please dive into the issues at https://github.com/trpc/v10-playground :)

**Reply by KATT** [reply]
Secondly, I won't ever allow you to define an `input` that isn't enforced by a validator :)
